Add APISIX post from APISIX Blog (#863)
oktadeveloper/okta-blog: _source/_posts/2021-09-18-apache-apisix-gateway.md

---
layout: blog_post
title: "Centralize Authentication at the Gateway with Apache APISIX and OpenID Connect"
author: xinxin-zhu
by: external-contributor
communities: [devops]
description: "Tutorial: Learn how to configure the Apache APISIX gateway to protect your APIs using OpenID Connect."
tags: [apisix, gateway, api, security, oauth2, oidc, oss]
tweets:
- "Use an API gateway to secure your applications. Learn how with Apache APISIX 🔒"

> - "An API gateway can make securing your APIs easy, follow along with this post‼️"

Reviewer: ```suggestion
- "The Apache APISIX Gateway with OpenID Connect can make it easier to secure your APIs. Here's the step-by-step 👣 ‼️" #opensource  
- Apache APISIX + Open ID Connect + Okta 👉  Learn how to replace traditional with centralized authentication at the gateway 👉 
```
emojifying

First, the user initiates a request, then the gateway itself takes charge of the user authentication process, interacting with the identity provider and sending them an authorization request. The identity provider returns user identity information (user info). After the gateway identifies the user, it forwards the user identity information (user info) to the services in a request header.

{% img blog/apache-apisix-gateway/2021-08-16-2.png alt:"Centralized Identity Authentication work flow" width:"600" %}{: .center-image }

Compared with the traditional authentication mode, centralized identity mode has the following advantages:

1. Centralized auth simplifies the application development process, and reduces the development application workload and maintenance costs by avoiding the need to repeat the development of authentication code for each application.
2. Centralized authentication mode improves business security. At the gateway level, it can intercept unauthenticated requests in time to protect back-end applications.

> ## What is OpenID Connect

Reviewer: ```suggestion
## What Is OpenID Connect
```

InfoQ title case

### OpenID Authentication Process

{% img blog/apache-apisix-gateway/2021-08-16-3.png alt:"OpenID Authentication Process" width:"600" %}{: .center-image }

1. APISIX initiates an authentication request to the Identity Provider.
2. The user logs in and authenticates on the Identity Provider.
3. The Identity Provider returns to APISIX with the Authorization Code.
4. APISIX requests the Identity Provider with the Code extracted from the request parameters.
5. The Identity Provider sends an answer message to APISIX with the ID Token and Access Token.
6. APISIX sends the Access Token to the Identity Provider's User Endpoint for authentication.
7. After passing the authentication, the User Endpoint sends the User Info to APISIX to complete  authentication.

> ## How to Configure Okta Authentication Using the Apache APISIX OpenID Connect Plug-in

Reviewer: ```suggestion
## How to Configure Okta Authentication Using the Apache APISIX OpenID Connect Plugin
```

- no hypen in plugin -

> Configuring Okta authentication using the Apache APISIX OpenID Connect plugin is a simple three-step process that allows you to switch from traditional authentication to centralized identity authentication mode. The following sections describe the steps to configure Okta authentication using the OpenID Connect plug-in for Apache APISIX.

Reviewer: ```suggestion
Configuring Okta authentication using the Apache APISIX OpenID Connect plugin is a three-step process that allows you to switch from traditional authentication to centralized identity authentication mode. The following sections describe the steps to configure Okta authentication using the OpenID Connect plugin for Apache APISIX.
```
removed "simple" because it's kind of judgmental, plus it's not really that simple....

    {% img blog/apache-apisix-gateway/2021-08-16-6.png alt:"Set the redirect URL for login and logout" width:"600" %}{: .center-image }   

4. After finishing the settings, click **Save** to save the changes.

    {% img blog/apache-apisix-gateway/2021-08-16-7.png alt:"save the changes" width:"600" %}{: .center-image }

5. Visit the General page of the application to obtain the following configuration, which is required to configure Apache APISIX OpenID Connect.

- **Client ID**: OAuth client ID, the application ID, which corresponds to `client_id` and `{YOUR_CLIENT_ID}` below.
- **Client secret**: OAuth client secret, the application key, which corresponds to `client_secret` and `{YOUR_CLIENT_SECRET}` below.
- **Okta domain**: The domain name used by the application, corresponding to `{YOUR_ISSUER}` below.

{% img blog/apache-apisix-gateway/2021-08-16-8.png alt:"obtain configuration" width:"600" %}{: .center-image }

### Step 2: Install Apache APISIX

> #### Install dependencies

Reviewer: ```suggestion
#### Install Dependencies
```
title case

#### Installation via source release

1. Create a directory named `apisix-2.7`.

    ```shell
    mkdir apisix-2.7
    ```

2. Download Apache APISIX Release source package.

    ```shell
    wget https://downloads.apache.org/apisix/2.7/apache-apisix-2.7-src.tgz
    ```

    You can also download the Apache APISIX release source package from the Apache APISIX website. The [Apache APISIX Official Website - Download Page](https://apisix.apache.org/downloads/) also provides source packages for Apache APISIX, APISIX Dashboard, and APISIX Ingress Controller.

3. Unzip the Apache APISIX Release source package.

    ```shell
    tar zxvf apache-apisix-2.7-src.tgz -C apisix-2.7
    ```

4. Install the runtime-dependent Lua libraries.

    ```shell
    # Switch to the apisix-2.7 directory
    cd apisix-2.7
    # Create dependencies
    make deps
    ```

#### Initializing Dependencies

Run the following command to initialize the NGINX configuration file and etcd.

```shell
# initialize NGINX config file and etcd
make init
```

### Step 3: Start Apache APISIX and Configure the Corresponding Route

1. Run the following command to start Apache APISIX.

    ```shell
    apisix start
    ```

> 2. Create a route and configure the OpenID Connect plugin. The following code example creates a route through the Apache APISIX Admin API, setting the upstream path to httpbin.org, a simple backend service for receiving and responding to requests. The following will use the get page of httpbin.org. Please refer to [httpbin get](http://httpbin.org/#/HTTP_Methods/get_get) for more information. For specific configuration items, please refer to the [Apache APISIX OpenID Connect Plugin](https://apisix.apache.org/docs/apisix/plugins/openid-connect/).

Reviewer: ```suggestion
2. Create a route and configure the OpenID Connect plugin. The following code example creates a route through the Apache APISIX Admin API, setting the upstream path to `httpbin.org`, a simple backend service for receiving and responding to requests. The following will use the get page of `httpbin.org`. Please refer to [`httpbin get`](http://httpbin.org/#/HTTP_Methods/get_get) for more information. For specific configuration items, please refer to the [Apache APISIX OpenID Connect Plugin](https://apisix.apache.org/docs/apisix/plugins/openid-connect/).
```
i think the urls will look better/scan better this way.

---
layout: blog_post
title: "Centralize Authentication at the Gateway with Apache APISIX and OpenID Connect"
author: xinxin-zhu
by: external-contributor
communities: [devops]
description: "Tutorial: Learn how to configure the Apache APISIX gateway to protect your APIs using OpenID Connect."
tags: [apisix, gateway, api, security, oauth2, oidc, oss]
tweets:
- "Use an API gateway to secure your applications. Learn how with Apache APISIX 🔒"
- "The Apache APISIX Gateway with OpenID Connect can make it easier to secure your APIs. Here's the step-by-step 👣 ‼️" #opensource  
- Apache APISIX + Open ID Connect + Okta 👉  Learn how to replace traditional with centralized authentication at the gateway 👉 
image: blog/apache-apisix-gateway/apisix-social.jpg
type: conversion
canonical-url: https://apisix.apache.org/blog/2021/08/16/Using-the-Apache-APISIX-OpenID-Connect-Plugin-for-Centralized-Authentication
---

> [Apache APISIX](https://github.com/apache/apisix) is a dynamic, real-time, high-performance API gateway, providing rich traffic management. The project offers load balancing, dynamic upstream, canary release, circuit breaking, authentication, observability, and many useful plugins. In addition, the gateway supports dynamic plugin changes along with hot-plugging. The OpenID Connect plugin for Apache APISIX allows users to replace traditional authentication mode with centralized identity authentication mode via OpenID Connect.
> 
> **Table of Contents**{: .hide }

Author: ```suggestion
[Apache APISIX](https://github.com/apache/apisix) is a dynamic, real-time, high-performance API gateway, providing rich traffic management. The project offers load balancing, dynamic upstream, canary release, circuit breaking, authentication, observability, and many useful plugins. In addition, the gateway supports dynamic plugin changes along with hot-plugging. The OpenID Connect plugin for Apache APISIX allows users to replace traditional authentication mode with centralized identity authentication mode via OpenID Connect.

> **NOTE:** This post originally appeared on the [Apache APISIX blog](https://apisix.apache.org/blog/2021/08/16/Using-the-Apache-APISIX-OpenID-Connect-Plugin-for-Centralized-Authentication).

**Table of Contents**{: .hide }
```
